core/src/main/java/io/grpc/internal/AbstractStream2.java

```diff
@@ -234,6 +247,19 @@ protected final void onStreamAllocated() {
     }
 
     /**
+     * Notify that the stream does not exist in a usable state any longer. This causes {@link
+     * AbstractStream2#isReady()} to return {@code false} from this point forward.
+     *
+     * <p>This does not generally need to be called explicitly by the transport, as it is handled
+     * implicitly by {@link AbstractClientStream2} and {@link AbstractServerStream}.
+     */
+    protected final void onStreamDeallocated() {
+      synchronized (onReadyLock) {
```

Comment:
Adding it causes failures, from `transportReportStatus()`. `transportReportStatus()` can process multiple times when `stopDelivery = true`. In general I find it can be useful for code processing error cases to be permissive to duplicate invocations (and this is called in error paths).